### PR TITLE
CLN: make submodules of pandas.util private

### PR DESCRIPTION
**Pull request by jreback**
xref #13634

### CODE REVIEW

File: pandas/util/hashing.py [hunk 1]
```diff
@@ -1,278 +1,18 @@
...
-# 16 byte long hashing key
-_default_hash_key = '0123456789123456'
+    def outer(t=t):
 
+        def wrapper(*args, **kwargs):
+            from pandas import util
+            warnings.warn("pandas.util.hashing is deprecated and will be "
+                          "removed in a future version, import "
+                          "from pandas.util",
+                          FutureWarning, stacklevel=3)
```
Comment (jorisvandenbossche) [inline pandas/util/hashing.py]: This is not needed I think, as I said above, see my recent PR here: https://github.com/pandas-dev/pandas/pull/16086/files, this was moved from pandas.tools to util.

File: asv_bench/benchmarks/attrs_caching.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 from .pandas_vb_common import *
-from pandas.util.decorators import cache_readonly
```
Comment (jorisvandenbossche) [inline asv_bench/benchmarks/attrs_caching.py]: You can do the public import here. But, you will also have to try the older import paths.

The hashing imports also can use an update I think

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good!

File: doc/source/whatsnew/v0.20.0.txt [hunk 3]
```diff
@@ -1238,10 +1238,10 @@ If indicated, a deprecation warning will be issued if you reference theses modul
     "pandas.types", "pandas.core.dtypes", ""
     "pandas.io.sas.saslib", "pandas.io.sas.libsas", ""
     "pandas._join", "pandas._libs.join", ""
-    "pandas._hash", "pandas.util.libhashing", ""
+    "pandas._hash", "pandas._libs.hashing", ""
     "pandas._period", "pandas._libs.period", ""
     "pandas._sparse", "pandas.core.sparse.libsparse", ""
-    "pandas._testing", "pandas.util.libtesting", ""
+    "pandas._testing", "pandas.util._testing", ""
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: no moving

File: doc/source/whatsnew/v0.20.0.txt [hunk 4]
```diff
@@ -1254,6 +1254,8 @@ these are now the public subpackages.
 
 - The function :func:`~pandas.api.types.union_categoricals` is now importable from ``pandas.api.types``, formerly from ``pandas.types.concat`` (:issue:`15998`)
 - The type import ``pandas.tslib.NaTType`` is deprecated and can be replaced by using ``type(pandas.NaT)`` (:issue:`16146`)
+- The public functions in ``pandas.util.hashing`` deprecated from that locations, but are now importable from ``pandas.util`` (:issue:`16223`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: done

File: asv_bench/benchmarks/attrs_caching.py (hunk 2 above)
Comment (jreback) [inline asv_bench/benchmarks/attrs_caching.py]: done

File: pandas/tools/hashing.py [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+for t in ['hash_pandas_object', 'hash_array']:
+
+    def outer(t=t):
+
+        def wrapper(*args, **kwargs):
+            from pandas import util
+            warnings.warn("pandas.tools.hashing is deprecated and will be "
+                          "removed in a future version, import "
+                          "from pandas.util",
+                          FutureWarning, stacklevel=3)
```
Comment (TomAugspurger) [inline pandas/tools/hashing.py]: Thoughts on making this one a DeprecationWarning? The only reason I ask is that dask did a release yesterday. Could do a point release fixing this if needed though.

**Review by TomAugspurger — APPROVED** [review]

File: pandas/tools/hashing.py (hunk 5 above)
Comment (jreback) [inline pandas/tools/hashing.py]: hmm, sure could do that.

**Comment by TomAugspurger** [comment]
Will have to update the test too.

**Comment by jreback** [comment]
nore that I didn't move ``pandas.util.testing`` as I suspect this is heavily used like ``pandas.core``. So will need to do something about that later.

**Comment by jreback** [comment]
> Will have to update the test too.

already did

**Comment by jorisvandenbossche** [comment]
Damn, this broke geopandas .. (we were apparently already using cache_readonly from utils.decorators)

And this will also break *released* version of dask (not master, as you already did a pr for that)

Any chance in doing a quick fix (0.20.1) before sending out the announcement?

**Comment by jreback** [comment]
does this not work with the latest released  of dask?

**Comment by jorisvandenbossche** [comment]
Yes, a clean install of pandas and dask currently does not work. 

Can we somehow 'hide' the conda packages for a moment? (I hided the release on pypi as there are no wheels yet)

**Comment by jreback** [comment]
we. an tag a post release
i think it's 0.20.0.post i think is valid

also fyi add to the release checklist
pypi upload is very last step (or ok if u immediately hide)
until conda and wheels are there we want to hide

**Pull request merged**